---
title: "Differential Gene Expression Analysis for 1"
author:
  - "Gan Lin, Genomics Facility, lgan@ukaachen.de"
  - "Dr. Chao-Chung Kuo, Genomics Facility, ckuo@ukaachen.de"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
library("heatmaply")
require(dplyr)
require(DT)
library(ggplot2)
library(DESeq2)
library("tximport")
library("readr")
library("tximportData")
library(stringr)
library(tidyr)
library('openxlsx')
library(AnnotationDbi, warn.conflicts = FALSE)
options(warn = -1)

DIR_base <- "GPM_DIR_BASE/"
DIR_project <- paste0(DIR_base, "analysis/")

source(paste0(DIR_project, "functions.R"))

load(paste0(DIR_project, "data.RData"))

if (organism == "hsapiens") {
  library(org.Hs.eg.db)
} else if (organism == "mmusculus") {
  library(org.Mm.eg.db)
}

Tag_this_analysis <- "M0_KO_vs_WT"
File_stats <- paste0('DGEA_',Tag_this_analysis,'_genes_stats.csv')
File_norm_expression <- paste0('DGEA_',Tag_this_analysis,'_norm_exp_quant.csv')
File_xlsx_res <- paste0('DGEA_',Tag_this_analysis,'_res.xlsx')

## Please assign the groups for comparison into labels_group variable and confirm the direction of comparison by level parameter in the function factor, here is just an example
# Manually define the grouping
labels_group <- c("control","control",
                  FALSE,FALSE,
                  FALSE,FALSE,
                  "knockdown","knockdown",
                  FALSE,FALSE,
                  FALSE,FALSE)
samples$group <- factor(labels_group, levels = c("control","knockdown"))
# Select by columns
# samples <- samples[samples$macrophages=="M0",]
# samples$group <- factor(samples$genotype, levels = c("WT","KO"))

```

### [Back to front page](Basic_analysis_RNAseq.html)

Differential expression analysis is done with DESEQ2 package in R [@love2014moderated] and follows the instruction for Salmon quantification ([Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)).

If there is a need to modify the figures to fit your particular need (style or color), additional service for customized visualization is needed. The main focus of this report is to present the analysis result to the clients. Any requests beyond this purpose is regarded another service. 

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
samples2 <- samples[!(is.na(samples$group)),]
deseq_output <- run_deseq_salmon(samples2, spikein=spikein_ERCC)
# samples2$sample <- samples2$rename
```

# Interactive Figures {.tabset}

## PCA

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
PCA_after_deseq2(deseq_output$norm_count, samples2)
```

## Volcano plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
volcano_plot(deseq_output$deseq2res)
```

## MA plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
MA_plot(deseq_output$deseq2res)
```

`r if(spikein_ERCC){"## MA plot of spike in"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, eval=spikein_ERCC}
MA_plot_ERCC(deseq_output$deseq2res_ERCC)
```

## Heatmap

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
heatmap_expression(deseq_output$deseq2res)
```


# Static Figures {.tabset}

## PCA

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
t <- t(deseq_output$norm_count[, c(-1,-2)])
prin_comp <- prcomp(t, rank. = 2)
components <- prin_comp[["x"]]
components <- data.frame(components)
components <- cbind(components, rownames(t))
labels <- samples2$sample
labels_group <- samples2$group

fig <- ggplot(components, aes(PC1, PC2, color=labels_group)) +
       geom_point(size=3) + theme_bw() + scale_color_brewer(palette = "Set1") + 
       labs(color = "Groups") + ggtitle("PCA") +
       theme(plot.title = element_text(hjust = 0.5))
fig
```

## Volcano plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
pal <- c("red", "royalblue")
pal <- setNames(pal, c("Sig. genes", "Non-sig."))
xmax <- max(abs(deseq_output$deseq2res$log2FoldChange)) * 1.1
ymax <- max(-log10(deseq_output$deseq2res$padj)[is.finite(-log10(deseq_output$deseq2res$padj))]) * 1.1
fig <- ggplot(deseq_output$deseq2res, aes(log2FoldChange, -log10(padj), color=sig)) +
       geom_point(size=1, alpha=0.2) + theme_bw() + scale_color_manual(values=pal) + 
       labs(color = "Groups") + ggtitle("Volcano plot") +
       xlab("Fold Change (log2)") + ylab("adjusted p-value (-log10)") +
       theme(plot.title = element_text(hjust = 0.5)) +
       xlim(-xmax, xmax) + ylim(0, ymax)
fig
```

## MA plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
pal <- c("red", "royalblue")
pal <- setNames(pal, c("Sig. genes", "Non-sig."))
ymax <- max(abs(deseq_output$deseq2res$log2FoldChange)) * 1.01

fig <- ggplot(deseq_output$deseq2res, aes(log2(baseMean), log2FoldChange, color=sig)) +
       geom_point(size=1, alpha=0.2) + theme_bw() + scale_color_manual(values=pal) + 
       labs(color = "Groups") + ggtitle("MA plot") +
       xlab("Expresion Mean (log2)") + ylab("Fold Change (log2)") +
       theme(plot.title = element_text(hjust = 0.5)) +
       ylim(-ymax, ymax)
fig
```

## Heatmap

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
margin_spacer <- function(x) {
  # where x is the column in your dataset
  left_length <- nchar(levels(factor(x)))[1]
  if (left_length > 8) {
    return((left_length - 8) * 4)
  }
  else
    return(0)
}

top1000 <- deseq_output$deseq2res[order(deseq_output$deseq2res$padj, decreasing = FALSE), ]
top1000 <- top1000[1:1000,]
samples_names <- colnames(top1000)[9:(dim(top1000)[2]-1)]
heatmap_t <- scale(log10(top1000[,9:(dim(top1000)[2]-1)]+1))
ord <- hclust( dist(heatmap_t, method = "euclidean"), method = "ward.D" )$order

heatmap_t <- cbind(top1000$gene_name, as.data.frame(heatmap_t))
colnames(heatmap_t)[1] <- "gene_name"
heatmap_t <- pivot_longer(heatmap_t, cols=2:(dim(heatmap_t)[2]), names_to="sample", values_to="Expression")
heatmap_t$gene_name <- factor( heatmap_t$gene_name, levels = top1000$gene_name[ord])
heatmap_t$sample <- factor( heatmap_t$sample, levels = samples_names)

fig <- ggplot(heatmap_t, aes(sample, gene_name, fill=Expression)) + 
       geom_tile() + ggtitle("Heatmap of top 1000 genes ranked by padj") +
       ylab("Genes") + xlab("Samples") + scale_fill_viridis() + 
       theme(plot.title = element_text(hjust = 0.5),
             axis.ticks.y = element_blank(),
             axis.text.y = element_blank(),
             axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
             plot.margin = margin(l = 0 + margin_spacer(heatmap_t$sample)))
fig
```

# Tables {.tabset}

## Sig. genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig)
```

## Up genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange > 0,])
```

## Down genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange < 0,])
```

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
# download_table()
cols <- c("ENSEMBL", "SYMBOL", "GENENAME")
ensemblkeys <- gsub("\\.[0-9]*$", "", deseq_output$deseq2res$gene_id)
if (organism == "hsapiens") {
  genenames <- select(org.Hs.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
} else if (organism == "mmusculus") {
  genenames <- select(org.Mm.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
}
genenames <- genenames[match(ensemblkeys, genenames$ENSEMBL), ]
deseq_output$deseq2res <- cbind(deseq_output$deseq2res, genenames)

write.csv(deseq_output$deseq2res,paste0(DIR_project, File_stats))
write.csv(deseq_output$norm_count,paste0(DIR_project, File_norm_expression))

wb <- createWorkbook()
addWorksheet(wb, "DGE_stats")
writeData(wb, "DGE_stats", deseq_output$deseq2res)
addWorksheet(wb, "norm_exp_quant")
writeData(wb, "norm_exp_quant", deseq_output$norm_count)
## Save workbook to working directory
saveWorkbook(wb, file = File_xlsx_res, overwrite = TRUE)
```


# Download statistics

* Download xlsx file for all results: [`r File_xlsx_res`](`r File_xlsx_res`) (However, CSV file is more recommended for robust analysis.)
* Download CSV file for the statistics of all valid genes in differential expression analysis (ENSEMBL ID, Gene Symbol and Gene name are included): [`r File_stats`](`r File_stats`)
* Download CSV file for the expression quantification of all genes: [`r File_norm_expression`](`r File_norm_expression`)


# References

<div id="refs"></div>


# R session information

```{r, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```