---
title: "Comparison of 1"
author:
  - "Gan Lin, Genomics Facility, lgan@ukaachen.de"
  - "Chao-Chung Kuo, Genomics Facility, ckuo@ukaachen.de"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
library("heatmaply")
require(dplyr)
require(DT)
library(ggplot2)
library(DESeq2)
library("tximport")
library("readr")
library("tximportData")
library(org.Hs.eg.db)
library(AnnotationDbi, warn.conflicts = FALSE)
options(warn = -1)

DIR_base <- "GPM_DIR_BASE/"
DIR_project <- paste0(DIR_base, "analysis/")

source(paste0(DIR_project, "functions.R"))

load(paste0(DIR_project, "data.RData"))

Tag_this_analysis <- "M0_WT_vs_KO"

## Please assign the groups for comparison into labels_group variable and confirm the direction of comparison by level parameter in the function factor, here is just an example
# Manually define the grouping
labels_group <- c("control","control",
                  FALSE,FALSE,
                  FALSE,FALSE,
                  "knockdown","knockdown",
                  FALSE,FALSE,
                  FALSE,FALSE)
samples$group <- factor(labels_group, levels = c("control","knockdown"))
# Select by columns
# samples <- samples[samples$macrophages=="M0",]
# samples$group <- factor(samples$genotype, levels = c("WT","KO"))

```

### [Back to front page](Basic_analysis_RNAseq.html)

Differential expression analysis is done with DESEQ2 package in R [@love2014moderated] and follows the instruction for Salmon quantification ([Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)).

# Figures {.tabset}

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
samples2 <- samples[!(is.na(samples$group)),]
deseq_output <- run_deseq_salmon(samples2)
# samples2$sample <- samples2$rename
```

## PCA

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
PCA_after_deseq2(deseq_output$norm_count, samples2)
```

## Volcano plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
volcano_plot(deseq_output$deseq2res)
```

## Heatmap

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
heatmap_expression(deseq_output$deseq2res)
```

# Tables {.tabset}

## Sig. genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig)
```

## Up genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange > 0,])
```

## Down genes

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange < 0,])
```

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
# download_table()
cols <- c("ENSEMBL", "SYMBOL", "GENENAME")
ensemblkeys <- gsub("\\.[0-9]*$", "", deseq_output$deseq2res$gene_id)
genenames <- select(org.Hs.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
genenames <- genenames[match(ensemblkeys, genenames$ENSEMBL), ]
deseq_output$deseq2res <- cbind(deseq_output$deseq2res, genenames)
write.csv(deseq_output$deseq2res,paste0(DIR_project, paste0('Diff_Exp_genes_stats_',Tag_this_analysis,'.csv')))
write.csv(deseq_output$norm_count,paste0(DIR_project, paste0('normalized_expression_quantification_',Tag_this_analysis,'.csv')))
```


# Download statistics

* Download CSV file for the statistics of all valid genes in differential expression analysis (ENSEMBL ID, Gene Symbol and Gene name are included): [`r paste0('Diff_Exp_genes_stats_',Tag_this_analysis,'.csv')`](`r paste0('Diff_Exp_genes_stats_',Tag_this_analysis,'.csv')`)
* Download CSV file for the expression quantification of all genes: [`r paste0('normalized_expression_quantification_',Tag_this_analysis,'.csv')`](`r paste0('normalized_expression_quantification_',Tag_this_analysis,'.csv')`)


# References

<div id="refs"></div>


# R session information

```{r, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```
