---
title: "Gene Ontology Analysis"
author:
GPM_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
Tag_this_analysis <- "GO"
pvalueCutoff_GO <- 0.05

load("data.RData")

library(gprofiler2)
library(clusterProfiler)
library(enrichplot)
library(DOSE)
library(curl)
library(DT)
library(kableExtra)

if (organism == "hsapiens") {
  library(org.Hs.eg.db)
} else if (organism == "mmusculus") {
  library(org.Mm.eg.db)
} else if (organism == "rnorvegicus") {
  library(org.Rn.eg.db)
}

table_gostres <- function(gostres_table) {
  tt <- as.data.frame(gostres_table)
  # For single query
  # tt <- tt[, c("source", "term_id", "term_name","significant","p_value","term_size","query_size","intersection_size",
  #              "precision","recall","effective_domain_size","source_order","parents")]
  # For multi queries with TRUE
  # tt <- tt[, c("source","term_id","term_name","p_values","significant","term_size", "query_sizes","intersection_sizes",
  #              "effective_domain_size","source_order","parents")]
  # For multi queries with FALSE
  tt <- tt[, c("source", "term_id", "term_name", "p_value", "significant", "term_size", "query",
               "query_size","intersection_size","precision","recall","effective_domain_size","source_order","parents")]
     
  DT::datatable( tt,
             extensions = c("FixedColumns"),
             filter = 'top',
             options = list( autoWidth = TRUE ,
                             dom = 'Blftip',
                             pageLength = 10,
                             searchHighlight = FALSE,
                             scrollX = TRUE,
                             fixedColumns = list(leftColumns = 3),
                             order = list(list(6))
                             ),
             class = c('compact cell-border stripe hover') ,
             rownames = FALSE) %>% formatRound(c("p_value", "precision", "recall"), 4)
}

```

### [Back to front page](Analysis_Report_RNAseq.html)

Interpretation of gene lists is a key step in numerous biological data analysis workflows, such as differential gene expression analysis and co-expression clustering of RNA-seq data. Usually this involves associating these gene lists with previous knowledge from well curated data sources of biological processes and pathways.

# Method

For gene ontology analysis, R library gprofiler2[@kolberg2020gprofiler2] was used. Using gost function, correction_method = 'FDR' and significant=TRUE. To minimize the enrichment of gene ontologies based on a small set of genes, we used three conditions for significance assessment: false discovery rate (FDR) < `r pvalueCutoff_GO`. We used Gene Ontology (GO or by branch GO\:MF, GO\:BP, GO\:CC), Kyoto Encyclopedia of Genes and Genomes (KEGG), WikiPathways (WP) as sources.

<!-- intersection size (intersection between gene set vs. a number of genes in a term) > 3; and precision (intersection size divided by gene set) >0.03 -->


```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
run_GO <- function(label) {
  GENE_LIST <- paste0("DGEA_", label, "_genes_stats.csv")
  # reading in input from deseq2
  df = read.csv(GENE_LIST, header=TRUE)
  df <- df[df$padj < CUTOFF_ADJP,]
  # de_gene_list <- gsub("\\.[0-9]*$", "", df$gene_id[(df$padj < pvalueCutoff)])
  up_gene_list <- gsub("\\.[0-9]*$", "", df$gene_id[(df$padj < CUTOFF_ADJP)&(df$log2FoldChange>0)])
  dn_gene_list <- gsub("\\.[0-9]*$", "", df$gene_id[(df$padj < CUTOFF_ADJP)&(df$log2FoldChange<0)])
  
  gostres <- gost(query = list("Down" = dn_gene_list,
                               "Up" = up_gene_list), 
                  multi_query = FALSE, evcodes = TRUE, 
                  correction_method = "fdr", user_threshold=pvalueCutoff_GO,
                  organism = organism, numeric_ns = "",
                  sources=c("GO", "KEGG", "WP"))
                  # sources=c("GO", "KEGG", "REAC", "MIRNA", "HP", "HPA", "WP"))
  # gostplot(gostres, capped = FALSE, interactive = TRUE)
  # p1 <- gostplot(gostres, capped = TRUE, interactive = FALSE)
  # publish_gostplot(p1, highlight_terms = gostres_de$result$term_id)
  write.table(gostres$result[1:11], file = paste0(Tag_this_analysis, "_", label,"_res.csv"))
  return(gostres)
}


fnames <- list.files(pattern = "*genes_stats.csv")
fnames <- fnames[fnames !="DGEA_All_samples_genes_stats.csv"]
labels <- gsub("DGEA_", "", fnames)
labels <- gsub("_genes_stats.csv", "", labels)

```

<!-- # GO analysis among treatments {.tabset} -->

<!-- ```{r,echo=FALSE, results='asis', warning=FALSE, message=FALSE} -->
<!-- for (lab in labels) { -->
<!--   cat("\n## ",  lab, "\n") -->
<!--   gostres <- run_GO(label=lab) -->
<!--   gostplot(gostres, capped = FALSE, interactive = TRUE) -->
<!--   cat("\n") -->
<!-- } -->
<!-- ``` -->

# WT {.tabset}

## IL10_vs_Control_in_WT

```{r, echo=FALSE, results='markup', warning=FALSE, message=FALSE}
gostres <- run_GO(label="IL10_vs_Control_in_WT")
gostplot(gostres, capped = FALSE, interactive = TRUE)
```

# Resources

* [Gene Ontology overview](http://geneontology.org/docs/ontology-documentation/)
* [g:Profiler](https://biit.cs.ut.ee/gprofiler/page/r)

# References

<div id="refs"></div>


# R session information

```{r, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```